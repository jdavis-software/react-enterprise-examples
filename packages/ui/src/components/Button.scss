@use '../tokens.scss' as tokens;

.ui-button {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  border: none;
  border-radius: var(--radius-2);
  padding: var(--space-2) var(--space-3);
  font: inherit;
  cursor: pointer;
  gap: var(--space-1);

  &--sm {
    padding: var(--space-1) var(--space-2);
    font-size: 0.875rem;
  }

  &--lg {
    padding: var(--space-3) var(--space-4);
    font-size: 1.125rem;
  }

  &--primary {
    background: var(--accent);
    color: var(--accent-contrast);
  }

  &--danger {
    background: var(--danger);
    color: var(--danger-contrast);
  }

  &--subtle {
    background: var(--surface-muted);
    color: var(--text-primary);
  }

  &--ghost {
    background: transparent;
    color: var(--text-primary);
  }

  &--loading {
    cursor: progress;
  }

  &:disabled {
    opacity: 0.6;
    cursor: not-allowed;
  }

  &:focus-visible {
    @include tokens.focus-ring;
  }
}
