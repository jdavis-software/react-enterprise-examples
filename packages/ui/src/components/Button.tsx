import { forwardRef, ButtonHTMLAttributes, ReactNode } from 'react';
import './Button.scss';

type Variants = 'primary' | 'danger' | 'subtle' | 'ghost';
type Sizes = 'sm' | 'md' | 'lg';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: Variants;
  size?: Sizes;
  isLoading?: boolean;
  children: ReactNode;
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      variant = 'primary',
      size = 'md',
      isLoading = false,
      disabled,
      children,
      className,
      ...props
    },
    ref
  ) => {
    const classes = [
      'ui-button',
      `ui-button--${variant}`,
      `ui-button--${size}`,
      isLoading ? 'ui-button--loading' : '',
      className
    ]
      .filter(Boolean)
      .join(' ');

    return (
      <button
        ref={ref}
        className={classes}
        disabled={disabled || undefined}
        aria-busy={isLoading || undefined}
        {...props}
      >
        {children}
      </button>
    );
  }
);

Button.displayName = 'Button';
