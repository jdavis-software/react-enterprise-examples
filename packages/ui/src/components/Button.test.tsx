import { render, screen, fireEvent } from '@testing-library/react';
import { describe, it, expect, vi } from 'vitest';
import { Button } from './Button';

describe('Button', () => {
  it('renders variants', () => {
    render(<Button variant="danger">Delete</Button>);
    expect(screen.getByRole('button')).toHaveClass('ui-button--danger');
  });

  it('variants snapshot', () => {
    const { container } = render(
      <div>
        <Button variant="primary">One</Button>
        <Button variant="danger">Two</Button>
      </div>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('fires onClick', () => {
    const onClick = vi.fn();
    render(<Button onClick={onClick}>Click</Button>);
    fireEvent.click(screen.getByText('Click'));
    expect(onClick).toHaveBeenCalled();
  });

  it('respects disabled', () => {
    const onClick = vi.fn();
    render(
      <Button disabled onClick={onClick}>
        Click
      </Button>
    );
    fireEvent.click(screen.getByText('Click'));
    expect(onClick).not.toHaveBeenCalled();
  });

  it('indicates busy state', () => {
    render(
      <Button isLoading>Save</Button>
    );
    expect(screen.getByRole('button')).toHaveAttribute('aria-busy', 'true');
  });
});
