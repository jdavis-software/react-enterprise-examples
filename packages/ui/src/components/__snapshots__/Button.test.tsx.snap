// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > variants snapshot 1`] = `
<div>
  <button
    class="ui-button ui-button--primary ui-button--md"
  >
    One
  </button>
  <button
    class="ui-button ui-button--danger ui-button--md"
  >
    Two
  </button>
</div>
`;
